File: MohaQuiz.Frontend/src/app/services/game-process.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { HubConnection } from '@microsoft/signalr';
import { BehaviorSubject, firstValueFrom } from 'rxjs';
import { environment } from 'src/environments/environment';
import { GameProcessStateModel } from '../models/game-process-state-model';

@Injectable({
  providedIn: 'root'
})
export class GameProcessService {

  BASE_URL = environment.apiUrl + 'gameprocess/'
  gameProcessState = new BehaviorSubject<GameProcessStateModel>({ gameName: '', roundNumber: 0, questionNumber: 0, isGameStarted: false, isScoring: false, isGameFinished: false })

  constructor(private http: HttpClient, public hc: HubConnection, private router: Router) {
    this.hc.start()
  }

  async startGame(gameName: string): Promise<any> {
    return await firstValueFrom(this.http.get(`${this.BASE_URL}startgame/${gameName}`))
  }

  async nextQuestion() {
    await firstValueFrom(this.http.get(`${this.BASE_URL}nextquestion`))
  }

  async startScoring() {
    await firstValueFrom(this.http.get(`${this.BASE_URL}startscoring`))
  }

  async stopScoring() {
    await firstValueFrom(this.http.get(`${this.BASE_URL}stopscoring`))
  }

  async nextRound() {
    await firstValueFrom(this.http.get(`${this.BASE_URL}nextround`))
  }

  async getActualGameProcess() {
    await firstValueFrom(this.http.get<GameProcessStateModel>(`${this.BASE_URL}getactualgameprocess`))
      .then(gp => this.gameProcessState.next(gp))
  }

  async resetGame() {
    await firstValueFrom(this.http.get(`${this.BASE_URL}resetgame`))
  }

  listenForStartGameMessage() {
    this.hc.on('StartGame', mes => {
      if (mes === 'StartGame') {
        this.router.navigate(['game'])
      }
    })
  }

  stopListenForStartGameMessage() {
    this.hc.off('StartGame')
  }

  listenForGameProcess() {
    this.hc.on('GetGameProcessState', gp => {
      this.gameProcessState.next(gp)
    })
  }

  stopListenForGameProcess() {
    this.hc.off('GetGameProcessState')
  }
}

File: MohaQuiz.Backend/Services/GameProcessService.cs
using Microsoft.AspNetCore.SignalR;
using MohaQuiz.Backend.Abstractions;
using MohaQuiz.Backend.Hubs;
using MohaQuiz.Backend.Models.DTOs;

namespace MohaQuiz.Backend.Services;

public class GameProcessService : IGameProcessService
{
    private static string _gameName = string.Empty;
    private static int _actualRoundNumber;
    private static int _actualQuestionNumber;
    private static bool _isGameStarted;
    private static bool _isScoring;
    private static bool _isGameFinished;
    private static readonly int _mAX_ROUND_NUMBER = 6;

    private readonly IHubContext<GameControlHub> _hubContext;

    public GameProcessService(IHubContext<GameControlHub> hubContext)
    {
        _hubContext = hubContext;
    }

    public async Task StartGame(string gameName)
    {
        _gameName = gameName;
        _actualRoundNumber = 1;
        _actualQuestionNumber = 0;
        _isGameStarted = true;

        await _hubContext.Clients.All.SendAsync("StartGame", "StartGame");
        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }

    public bool IsGameStarted()
    {
        return _isGameStarted;
    }

    public async Task NextQuestion()
    {
        _actualQuestionNumber++;

        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }

    public async Task StartScoring()
    {
        _isScoring = true;

        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }


    public async Task StopScoring()
    {
        _isScoring = false;
        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }

    public async Task NextRound()
    {
        _actualRoundNumber++;
        _actualQuestionNumber = 0;
        _isScoring = false;

        if(_actualRoundNumber > _mAX_ROUND_NUMBER)
        {
            _isGameFinished = true;
        }

        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }

    public async Task ResetGame()
    {
        _gameName = string.Empty;
        _actualRoundNumber = 0;
        _actualQuestionNumber = 0;
        _isGameStarted = false;
        _isScoring = false;
        _isGameFinished = false;

        await _hubContext.Clients.All.SendAsync("GetGameProcessState", GetActualGameProcess());
    }

    public static GameProcessStateDTO GetActualGameProcess()
    {
        return new GameProcessStateDTO() { GameName = _gameName,
                                           RoundNumber = _actualRoundNumber,
                                           QuestionNumber = _actualQuestionNumber,
                                           IsGameStarted = _isGameStarted,
                                           IsScoring = _isScoring,
                                           IsGameFinished = _isGameFinished};
    }

}
